### PR TITLE
linter: Added saving of constant value in ConstantInfo #583

### PR DESCRIPTION
**Pull request by i582**
solve #583

### CODE REVIEW

File: src/linter/block.go [hunk 1]
```diff
@@ -1361,6 +1362,27 @@ func (b *BlockWalker) handleArrayItems(arr node.Node, items []*expr.ArrayItem) b
...
+
+			key = info.Value.Value
+
+			if info.Value.Type == meta.Float {
+				val, err := strconv.ParseFloat(key, 64)
+				if err != nil {
+					continue
+				}
+				value := int64(math.Floor(val))
+				key = fmt.Sprint(value)
```
Comment (quasilyte) [inline src/linter/block.go:1381]: You might want to test this code path.

From the PHP manual:
> Floats are also cast to integers, which means that the fractional part will be truncated. E.g. the key 8.7 will actually be stored under 8. 

This gives a hint that values are "truncated", not floor'ed.

For some inputs, it behaves the same, but:
```
int(floor(-1.51)) => int(-2.0) => -2
int(-1.51) => -1
```
PHP will do the latter:
```php
var_dump([-1.5 => 1]);
```
```
array(1) {
  [-1]=>
  int(1)
}
```
I would suggest skipping the floor step.
It's also a good idea to add tests that check this corner case behavior for `-1.51`.

File: src/linter/block.go [hunk 2]
```diff
@@ -1361,6 +1362,27 @@ func (b *BlockWalker) handleArrayItems(arr node.Node, items []*expr.ArrayItem) b
...
+			} else if info.Value.Type == meta.Undefined {
```
Comment (quasilyte) [inline src/linter/block.go:1382]: I would suggest moving this check closer to the !ok, so we get guarding clauses above the code that handles correct known values.

File: src/linter/block.go [hunk 3]
```diff
@@ -210,7 +211,7 @@ func (b *BlockWalker) checkBinaryDupArgs(n, left, right node.Node) {
 	if !b.sideEffectFree(left) || !b.sideEffectFree(right) {
 		return
 	}
-	if astutil.NodeEqual(left, right) {
+	if b.NodeEqual(left, right) {
```
Comment (quasilyte) [inline src/linter/block.go:214]: This may now match `Foo == Bar` as duplicated if their values are identical.
It might look like a false positive for a user.
Perhaps we need to adjust the report message so it looks less broken for this kind of situation.

File: src/linter/root.go [hunk 4]
```diff
@@ -857,11 +857,14 @@ func (d *RootWalker) enterClassConstList(s *stmt.ClassConstList) bool {
 		d.checkCommentMisspellings(c, c.PhpDocComment)
 		typ := solver.ExprTypeLocal(d.scope(), d.ctx.st, c.Expr)
 
+		value, _ := solver.GetConstantValue(c, d.ctx.st)
```
Comment (quasilyte) [inline src/linter/root.go:860]: GetConstantValue may call GetConstant and this creates a bad situation where the order of the file traversal matters. We try to avoid that.

In general, you can't use solver methods that rely on the metadata during the indexing.

I would limit GetConstantValue to AST nodes and avoid querying the metadata during the indexing.

**Comment by quasilyte** [comment]
I would also suggest trying to write more detailed commit messages.

This might be helpful:
https://golang.org/doc/contribute.html#commit_messages
